erpnext/controllers/selling_controller.py

```diff
@@ -548,3 +548,48 @@ def update_project_billing_and_sales(self):
 			doc.set_gross_margin(update=True)
 			doc.set_status(update=True)
 			doc.notify_update()
+
+	@frappe.whitelist()
+	def set_rate_as_cost(self):
+		for item in self.items:
+			if item.item_code:
+				cost_rate = get_rate_as_cost(item.item_code, item.warehouse, item.batch_no, item.serial_no,
+					self.posting_date, self.posting_time, item.qty, self.doctype, self.name,
+					self.company, item.delivery_note, item.delivery_note_item)
+				item.rate = cost_rate
+				item.discount_percentage = 0
+				item.margin_rate_or_amount = 0
+
+		self.calculate_taxes_and_totals()
+
+
+@frappe.whitelist()
+def get_rate_as_cost(item_code, warehouse, batch_no, serial_no, posting_date, posting_time, qty, voucher_type,
+		voucher_no, company, delivery_note=None, delivery_note_item=None):
+	qty = -1 * qty
+	if delivery_note and delivery_note_item:
+		sum_of_stock_value_difference = frappe.db.sql("""
+			SELECT SUM(stock_value_difference)
+			FROM `tabStock Ledger Entry` 
+			WHERE voucher_type = 'Delivery Note' AND voucher_no = %s AND voucher_detail_no = %s
+		""", (delivery_note, delivery_note_item))
+
+		cost_rate = flt(sum_of_stock_value_difference[0][0]) / qty
```

Comment:
A SQL query can give you a null. This line can give you an error for accessing a None value. This happens mostly when there are no rows that match the criteria in the SQL.

 _Always_ check for division by zero whenever you write a division. Qty can be (in some cases) be 0 which would lead to a Division By Zero error.

- Check for null before accessing `sum_of_stock_value_difference[0][0]`
- Inline check for falsy qty